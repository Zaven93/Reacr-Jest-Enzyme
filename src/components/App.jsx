import React, {Component} from "react"
import {Button} from "react-bootstrap"

import {max_number} from "../helper"
import Gift from "./Gift"

class App extends Component {
    constructor(){
        super()

        this.state = {
            gifts: []
        }
    }

    addGift = () => {
        const {gifts} = this.state

        const ids = this.state.gifts.map(gift => gift.id)

        const max_id = max_number(ids)

        gifts.push({id: max_id + 1})

        this.setState({gifts})
    }

    removeGift = (id) => {
        const filteredGifts = this.state.gifts.filter(gift => gift.id !== id)

        this.setState({gifts: filteredGifts})
    }

    render(){
        return (
            <div>
                <h2>Gift Giver</h2>
                <Button className="btn-add" onClick={this.addGift}>Add a Gift</Button>
                <div className="gift-list">
                    {
                        this.state.gifts.map(gift => (
                            <Gift gift removeGift={this.removeGift} key={gift.id} />
                        ))
                    }
                </div>
            </div>
        )
    }
}

export default App