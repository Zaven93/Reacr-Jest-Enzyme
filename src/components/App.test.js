import React from "react"
import Enzyme, {shallow} from "enzyme"
import Adapter from "enzyme-adapter-react-16"
import App from "./App"

Enzyme.configure({adapter: new Adapter()})

describe("App", () => {
    const app = shallow(<App />)

    it("renders correctly", () => {
        expect(app).toMatchSnapshot()
    })

    it("initializes the `state` with an empty list of gifts", () => {
        expect(app.state().gifts).toEqual([])
    })

    describe("when `Add Gift` button clicked", () => {
        const id = 1

        beforeEach(() => {
            app.find(".btn-add").simulate("click")
        })

        afterEach(() => {
            app.setState({gifts: []})
        })

        it("adding new gift to the state when add gift button is clicked", () => {
            expect(app.state().gifts).toEqual([{id}])
        })
    
        it("renders a new gift item when the add gift button is clicked", () => {
            expect(app.find(".gift-list").children().length).toEqual(id)
        })

        it("creates a Gift component", () => {
            expect(app.find("Gift").exists()).toBe(true)
        })

        describe("and the user wants to remove the gift", () => {
            beforeEach(() => {
                app.instance().removeGift(id)
            })
    
            it("removes the gift from state", () => {
                expect(app.state().gifts).toEqual([])
            })
        })
    })
})