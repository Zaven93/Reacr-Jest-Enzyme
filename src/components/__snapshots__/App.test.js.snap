// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <h2>
    Gift Giver
  </h2>
  <Button
    active={false}
    className="btn-add"
    disabled={false}
    onClick={[Function]}
    variant="primary"
  >
    Add a Gift
  </Button>
  <div
    className="gift-list"
  />
</div>
`;
